datafusion/optimizer/src/common_subexpr_eliminate.rs

```diff
@@ -275,68 +367,93 @@ impl CommonSubexprEliminate {
         config: &dyn OptimizerConfig,
     ) -> Result<Transformed<LogicalPlan>> {
         // collect all window expressions from any number of LogicalPlanWindow
-        let ConsecutiveWindowExprs {
-            window_exprs,
-            arrays_per_window,
-            expr_stats,
-            plan,
-        } = ConsecutiveWindowExprs::try_new(window)?;
+        let (mut window_exprs, mut window_schemas, mut plan) =
+            get_consecutive_window_exprs(window);
 
-        let arrays_per_window = arrays_per_window
+        let mut found_common = false;
+        let mut expr_stats = ExprStats::new();
+        let arrays_per_window = window_exprs
             .iter()
-            .map(|arrays| arrays.as_slice())
-            .collect::<Vec<_>>();
+            .map(|window_expr| {
+                self.to_arrays(window_expr, &mut expr_stats, ExprMask::Normal)
+                    .map(|(fc, id_arrays)| {
+                        found_common |= fc;
 
-        // save the original names
-        let name_preserver = NamePreserver::new(&plan);
-        let mut saved_names = window_exprs
-            .iter()
-            .map(|exprs| {
-                exprs
-                    .iter()
-                    .map(|expr| name_preserver.save(expr))
-                    .collect::<Result<Vec<_>>>()
+                        id_arrays
+                    })
             })
             .collect::<Result<Vec<_>>>()?;
 
-        assert_eq!(window_exprs.len(), arrays_per_window.len());
-        let num_window_exprs = window_exprs.len();
-        let rewritten_window_exprs = self.rewrite_expr(
-            window_exprs,
-            &arrays_per_window,
-            plan,
-            &expr_stats,
-            config,
-        )?;
-        let transformed = rewritten_window_exprs.transformed;
+        if found_common {
+            // save the original names
+            let name_preserver = NamePreserver::new(&plan);
+            let mut saved_names = window_exprs
+                .iter()
+                .map(|exprs| {
+                    exprs
+                        .iter()
+                        .map(|expr| name_preserver.save(expr))
+                        .collect::<Result<Vec<_>>>()
+                })
+                .collect::<Result<Vec<_>>>()?;
 
-        let (mut new_expr, new_input) = rewritten_window_exprs.data;
+            assert_eq!(window_exprs.len(), arrays_per_window.len());
+            let num_window_exprs = window_exprs.len();
+            let rewritten_window_exprs = self.rewrite_expr(
+                window_exprs.clone(),
```

Comment:
This extra clone might look bad at first, but technically this is due to we use references to original expressions in `Identifier`s so we have to keep the original expressions intact.
- The reason why this is not worse then before this PR is because previously the `Identifiers` were stringified expression trees (so they were kind of clones of the original expressions).
- The reason why this can be better then before this PR is because we introduce the `found_common` flag above that allows skipping cloning when it make no sense for the second, rewriting traversal.

Reply:
```suggestion
                // Must clone as Identifiers use references to original expressions
                // so we have to keep the original expressions intact.
                window_exprs.clone(),
```

in terms of the cost of the extra clone, I think the performance results speak for themselves